src/pybamm/solvers/c_solvers/idaklu/IDAKLUSolverOpenMP.inl

```diff
@@ -289,9 +295,13 @@ void IDAKLUSolverOpenMP<ExprSet>::Initialize() {
   realtype *id_val;
   id_val = N_VGetArrayPointer(id);
 
+  // Determine if the system is an ODE
+  is_ODE = number_of_states > 0;
   int ii;
   for (ii = 0; ii < number_of_states; ii++) {
-    id_val[ii] = id_np_val[ii];
+    const bool id_i = id_np_val[ii];
+    id_val[ii] = id_i;
+    is_ODE &= id_i == 1;
```

Comment:
If `id_i` is a bool, then the `id_i == 1` should be redundant

Reply:
This bit is a confusing jumble of types, these are not really bool, but are floats indicating if the state is differential (1.0) or algebraic (0.0), this is just the format that sundials expects. I think introducing a bool here is just confusing an already confusing sundials thing (e.g. they never say in the docs what happens to values that are not 0 or 1!), perhaps:
```c++
id_val[ii] = id_np_val[ii];
is_ODE &= id_val[ii] == 1.0
```

Reply:
Float comparisons are not reliable either. Float equality checks are not good to have in the code

Reply:
@martinjrobins good point! I'll make this change

Reply:
I made this an inequalty check to handle minor floating point arithmetic issues

Reply:
@kratman: totally agree! but this is how sundials docs define this vector (as a vector of floats, 1.0 = differential and 0.0 == algebraic) so I think we have to follow this definition, no matter how strange it is.

Reply:
saying that, its not a big deal so happy with `is_ODE &= id_val[ii] > SMALL_NUMBER`, as long as the type conversions are removed. it shouldn't make any difference to the result

Reply:
Yeah sorry for the confusion, I was talking to Marc separately removing the type conversion was already part of it